### PR TITLE
feat: configurable loading of event bus consumer

### PR DESCRIPTION
**Pull request by navinkarkera**
**Description:** This defines an API for IDAs to use when consuming events that allows the choice of Event Bus implementation to be configured rather than written in code. 

It also includes a management command to start a worker and allows to pass in extra arguments for the underlying implementation of event bus.

**ISSUE:** #147 

**Dependencies:** None

**Merge deadline:** List merge deadline (if any)

**Testing instructions:**

Please refer to https://github.com/openedx/event-bus-redis/pull/13 for testing instructions.

**Reviewers:**
- [ ] @kaustavb12
- [ ] @bmtcril 

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPI after tag-triggered build is
      finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:** The management command implement in [event-bus-kafka takes in an argument](https://github.com/openedx/event-bus-kafka/blob/3f0804234449bd1393198896db104f1f93743b7b/edx_event_bus_kafka/internal/consumer.py#L646-L649) called `offset_time` and calls a different method i.e. calls `reset_offsets_and_sleep_indefinitely` instead of `consume_indefinitely`. If this MR is accepted, we will need to change the implementation and set the argument `offset_time` in the class and move this [logic](https://github.com/openedx/event-bus-kafka/blob/3f0804234449bd1393198896db104f1f93743b7b/edx_event_bus_kafka/internal/consumer.py#L646-L649) inside `consume_indefinitely`.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @navinkarkera!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

**Review by kaustavb12 — APPROVED** [review]
:+1:

- [X] I tested this: Followed testing instructions
- [x] I read through the code
- [ ] I checked for accessibility issues
- [ ] Includes documentation
- [ ] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.

File: openedx_events/event_bus/tests/test_loader.py
Comment (kaustavb12) [inline openedx_events/event_bus/tests/test_loader.py:89]: @navinkarkera This test case seems to be failing in my local env with the following error. Please take a look at this.

```
>       assert [str(w.message) for w in caught_warnings] == warning_messages
E       assert ['Failed to l... be inactive'] == ['Failed to l... be inactive']
E         At index 0 diff: "Failed to load <class 'dict'> from setting EB_LOAD_PATH: TypeError('dict() argument after * must be an iterable, not int'); component will be inactive" != "Failed to load <class 'dict'> from setting EB_LOAD_PATH: TypeError('type object argument after * must be an iterable, not int'); component will be inactive"
E         Use -v to get more diff

```

Comment (navinkarkera) [reply openedx_events/event_bus/tests/test_loader.py:89]: @kaustavb12 Weird, it is working for me. Can you make sure your python version is 3.8 and the requirements are updated. (remove codecov from the requirements/dev.txt file manually).

Comment (kaustavb12) [reply openedx_events/event_bus/tests/test_loader.py:89]: @navinkarkera Actually I was using python version 3.10. Switching to 3.8, the tests ran successfully as it is.

File: openedx_events/apps.py
Comment (mariajgrimaldi) [inline openedx_events/apps.py:13]: Out of curiosity, why do we need this file now?

File: openedx_events/event_bus/__init__.py
Comment (mariajgrimaldi) [inline openedx_events/event_bus/__init__.py:159]: what are we documenting here?

File: openedx_events/event_bus/tests/test_loader.py
Comment (mariajgrimaldi) [inline openedx_events/event_bus/tests/test_loader.py:120]: is there a way to mock the not default behavior?

File: openedx_events/management/commands/consume_events.py
Comment (mariajgrimaldi) [inline openedx_events/management/commands/consume_events.py:25]:
```suggestion
        python manage.py consume_events -t user-login -g user-activity-service \
```
And so on

File: openedx_events/management/commands/consume_events.py
Comment (mariajgrimaldi) [inline openedx_events/management/commands/consume_events.py:28]: can we not be technology specific? or is it needed?

File: openedx_events/management/commands/consume_events.py
Comment (mariajgrimaldi) [inline openedx_events/management/commands/consume_events.py:48]: not needed extra blank line

File: openedx_events/management/commands/consume_events.py
Comment (mariajgrimaldi) [inline openedx_events/management/commands/consume_events.py:74]: what is `extra`? can we be a bit more specific for those reading the code?

File: openedx_events/tests/test_consume_events_command.py
Comment (mariajgrimaldi) [inline openedx_events/tests/test_consume_events_command.py:25]: can we add docstrings here? you can follow the format of tests like this one: https://github.com/openedx/openedx-events/blob/main/openedx_events/tests/test_tooling.py#L94-L99

**Review by mariajgrimaldi — CHANGES_REQUESTED** [review]
Hello there! I left a few comments regarding the style of the implementation :)!

**Comment by mariajgrimaldi** [comment]
I believe you folks could have more insight here than I currently have: @robrap @rgraber

**Comment by robrap** [comment]
Thanks @mariajgrimaldi. We can also give this a review when @navinkarkera let’s us know it’s ready.

File: openedx_events/apps.py
Comment (navinkarkera) [reply openedx_events/apps.py:13]: This converts openedx_events to a django app and allows the host to run `consume_events` management command if they include `openedx_events` in their list of `INSTALLED_APPS`.

File: openedx_events/event_bus/__init__.py
Comment (navinkarkera) [reply openedx_events/event_bus/__init__.py:159]: `EVENT_BUS_CONSUMER` at like 178 similar to `EVENT_BUS_PRODUCER` in line 105.

File: openedx_events/management/commands/consume_events.py
Comment (navinkarkera) [reply openedx_events/management/commands/consume_events.py:28]: @mariajgrimaldi there are additional features that we can use depending on the underlying tech, for example kafka has something related to resetting offset (not exactly sure how it works) and redis streams allows to start consuming from a specific id or read all the consumer groups unread msgs, claim messages from other consumers etc. The default behavior is same i.e. read new msgs only.

File: openedx_events/event_bus/tests/test_loader.py
Comment (navinkarkera) [reply openedx_events/event_bus/tests/test_loader.py:120]: @mariajgrimaldi The loader is already tested [above](https://github.com/openedx/openedx-events/pull/200/files#diff-d3aa2802dab4bb0b7464d190bc4f0c0b07655847877298bd51ea6ada75051a0fR39-R45) and we don't need to test `consume_indefinitely` here.

File: openedx_events/management/commands/consume_events.py
Comment (navinkarkera) [reply openedx_events/management/commands/consume_events.py:74]: Added a comment.

**Comment by navinkarkera** [comment]
@mariajgrimaldi Thanks for the comments, I have addressed them.

@robrap @rgraber It is ready for your review.

cc: @timmc-edx Adding you here, since all this is based on your work.

File: openedx_events/event_bus/__init__.py
Comment (timmc-edx) [inline openedx_events/event_bus/__init__.py:160]: These kwargs are here to provide additional implementation-specific configuration, right? In the Kafka implementation we just used Django settings namespaced under `EVENT_BUS_KAFKA_` for that purpose, and I would prefer that if possible. Here's an example of that:

- `create_producer` reads configuration before creating an `EventBusProducer`: https://github.com/openedx/event-bus-kafka/blob/main/edx_event_bus_kafka/internal/producer.py#L339
- Configuration is loaded from Django settings: https://github.com/openedx/event-bus-kafka/blob/main/edx_event_bus_kafka/internal/config.py#L72

Is there some other constraint that makes this JSON-and-kwargs preferable here?

Comment (navinkarkera) [reply openedx_events/event_bus/__init__.py:160]: @timmc-edx The reason was that the values for kwargs can change case by case and they are meant to be used in  special cases only, for example, if the users want to process previous events from a specific point or they want to run multiple consumers in a group so they need to pass in `--consumer_name` for each worker which should be unique in the group. Adding these in the settings means they will need to update the settings each time they want to start a worker.

File: manage.py
Comment (navinkarkera) [reply manage.py:19]: Yep, I added it after looking at few examples in openedx org. I'll remove it.

File: openedx_events/event_bus/__init__.py
Comment (timmc-edx) [reply openedx_events/event_bus/__init__.py:160]: That's actually how we've been handling it with Kafka -- change the settings, which automatically causes a redeploy of the workers on kubernetes, and then change them back again. But I can see how that approach might not be desirable for all deployers.

**Review by timmc-edx — APPROVED** [review]
This looks good to me, but I wrote some chunks of it, so I'd like to hear what Robert and/or Becca have to say. :-)

**Review by robrap — COMMENTED** [review]
Some high-level comments:
1. Do you intend to update https://openedx.atlassian.net/wiki/spaces/AC/pages/3508699151/How+to+start+using+the+Event+Bus right after merge, or do we want any local docs? Note that not being able to keep doc changes versioned with the code changes is not great. It might be time to move this how-to into the repo, even if we don't do it for this PR.
2. Is this management command replacing a management command in `edx_event_bus_kafka`? @timmc-edx: I assume we need a ticket for this work? 
3. @timmc-edx: Can you (or did you) address the author concern from the PR description around `reset_offsets_and_sleep_indefinitely`?

**Comment by navinkarkera** [comment]
@robrap I do have a ticket in my pipeline to update all related docs. I can also move it to the repo if it helps.

>Is this management command replacing a management command in edx_event_bus_kafka?

If the approach looks good, I can update `event_bus_kafka`.

>Can you (or did you) address the author concern from the PR description around reset_offsets_and_sleep_indefinitely?

Same, I can do it.

File: openedx_events/event_bus/__init__.py
Comment (timmc-edx) [reply openedx_events/event_bus/__init__.py:159]: `EVENT_BUS_PRODUCER` works the same way (accepting a dotted module string to a Callable) so I think we should stick to a parallel name.

File: openedx_events/event_bus/__init__.py
Comment (timmc-edx) [reply openedx_events/event_bus/__init__.py:160]: In our case, the Django settings and the command line arguments are in the same file as each other, so it doesn't really matter much which mechanism we use. I don't know how many people would be using this kind of setup, though.

If Redis needs individual names for the consumers in a group, I guess it seems more likely that you'd have a list of individual workers hardcoded into your infrastructure, each with its own command line. It might be harder to give each one its own Django settings. (That would be true for our infrastructure, in fact.)

So I'm fine with the `--extra` arg. (Well, I feel like there could be a better name for it, but I can't actually _come up with_ anything better, so I really can't complain there!)

Comment (timmc-edx) [reply openedx_events/event_bus/__init__.py:160]: Oh! I was actually totally wrong about how we were passing offset time to Kafka. We're using a CLI argument to the management command, just like you noted in the PR description. :-) So I guess we'd make use of this as well!

**Comment by timmc-edx** [comment]
@robrap Yes, this would replace the management command we have in event-bus-kafka. I have [an old branch that would complement this PR](https://github.com/openedx/event-bus-kafka/compare/main...timmc/consumer-api) but even though it's pretty out of date I think the spirit of it is still accurate. We would be able to cut over from the existing one to the new one without too much trouble, I think. There's an [existing closed ticket](https://github.com/openedx/event-bus-kafka/issues/19) for the Kafka impl using a consumer API like this, so we should probably reopen and rework it.

The issue of `reset_offsets_and_sleep_indefinitely` shouldn't be an issue -- we'll need to pass an `--extra` and then in our implementation, parse the JSON to get any offset info and pass that to that function.

File: openedx_events/event_bus/tests/test_loader.py
Comment (mariajgrimaldi) [reply openedx_events/event_bus/tests/test_loader.py:120]: good! thanks

File: openedx_events/management/commands/consume_events.py
Comment (mariajgrimaldi) [reply openedx_events/management/commands/consume_events.py:74]: thank you!

File: openedx_events/management/commands/consume_events.py
Comment (mariajgrimaldi) [reply openedx_events/management/commands/consume_events.py:28]: I get it now! Thank you

File: openedx_events/apps.py
Comment (mariajgrimaldi) [reply openedx_events/apps.py:13]: Do you folks think of any downside of turning the library into a django application? I'm just a bit worried about the implications (I haven't found any downside though). Let me know!

File: openedx_events/event_bus/__init__.py
Comment (navinkarkera) [reply openedx_events/event_bus/__init__.py:160]: Cool, I'll leave this as is then.

**Comment by navinkarkera** [comment]
> The issue of reset_offsets_and_sleep_indefinitely shouldn't be an issue -- we'll need to pass an --extra and then in our implementation, parse the JSON to get any offset info and pass that to that function.

@timmc-edx we just need to make constructor of kafka consumer accept `offset` as a kwarg and update `consume_indefinitely` to run reset offset if it is set, the parsing of json happens here.

File: openedx_events/event_bus/__init__.py
Comment (navinkarkera) [reply openedx_events/event_bus/__init__.py:159]: Updated!

Comment (navinkarkera) [reply openedx_events/event_bus/__init__.py:159]: Btw, updating `EVENT_BUS_PRODUCER` will require changes in [edx-platform](https://github.com/search?q=repo%3Aopenedx%2Fedx-platform+EVENT_BUS_PRODUCER&type=code) as well.

**Review by mariajgrimaldi — APPROVED** [review]

File: openedx_events/event_bus/__init__.py
Comment (timmc-edx) [reply openedx_events/event_bus/__init__.py:159]: @robrap How strongly do you feel about a name including `CALLABLE`? I had been hoping to keep it short.

Comment (robrap) [reply openedx_events/event_bus/__init__.py:159]: We can leave as-is.

Comment (navinkarkera) [reply openedx_events/event_bus/__init__.py:159]: Reverted.

File: openedx_events/management/commands/consume_events.py
Comment (timmc-edx) [inline openedx_events/management/commands/consume_events.py:65]: I think it would be good to clarify that this JSON has to decode to a dict.
```suggestion
            help='JSON object to pass additional arguments to the consumer.'
```

**Comment by timmc-edx** [comment]
(I've looked at all the related PRs by now as well.)

**Comment by navinkarkera** [comment]
@timmc-edx Thanks. Someone with write access needs to merge this.
cc: @rgraber @bmtcril

File: CHANGELOG.rst
Comment (timmc-edx) [inline CHANGELOG.rst:21]: One last change, now that it's ready to go.
```suggestion
[7.1.0] - 2023-05-03
```

Comment (timmc-edx) [reply CHANGELOG.rst:21]: I'll go ahead and merge, and will update this in a separate PR. (I don't have permissions to update this branch.)

**Comment by openedx-webhooks** [comment]
@navinkarkera 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/openedx-events&entry.867055334=https://github.com/openedx/openedx-events/pull/200&entry.1484655318=https://github.com/navinkarkera&entry.752974735=2023-04-12+15:25&entry.1917517419=2023-05-03+19:19&entry.2133058324=Yes) so we can improve your experience in the future.

File: CHANGELOG.rst
Comment (navinkarkera) [reply CHANGELOG.rst:21]: @timmc-edx Thanks!

**Pull request merged**
